core/embed/rust/src/ui/model_tr/bootloader/mod.rs

```diff
@@ -0,0 +1,405 @@
+use crate::{
+    trezorhal::io::io_button_read,
+    ui::{
+        component::{Component, Never},
+        display::{self, Font},
+        geometry::Point,
+        model_tr::constant,
+    },
+};
+
+mod confirm;
+mod intro;
+mod menu;
+mod theme;
+mod title;
+
+use crate::ui::{
+    component::{
+        text::paragraphs::{Paragraph, ParagraphVecShort, Paragraphs, VecExt},
+        Event, EventCtx,
+    },
+    constant::{screen, BACKLIGHT_NORMAL, WIDTH},
+    display::{fade_backlight_duration, Color, TextOverlay},
+    event::ButtonEvent,
+    geometry::{LinearPlacement, Offset, Rect},
+    model_tr::{
+        bootloader::{
+            confirm::Confirm,
+            intro::Intro,
+            menu::Menu,
+            theme::{bld_button_cancel, bld_button_default, BLD_BG, BLD_FG},
+        },
+        component::{Button, ButtonPos, ResultScreen},
+        theme::LOGO_EMPTY,
+    },
+    util::{from_c_array, from_c_str},
+};
+
+const SCREEN_ADJ: Rect = screen().split_top(64).0;
+
+pub trait ReturnToC {
+    fn return_to_c(self) -> u32;
+}
+
+impl ReturnToC for Never {
+    fn return_to_c(self) -> u32 {
+        unreachable!()
+    }
+}
+
+impl ReturnToC for () {
+    fn return_to_c(self) -> u32 {
+        0
+    }
+}
+
+fn button_eval() -> Option<ButtonEvent> {
+    let event = io_button_read();
+    if event == 0 {
+        return None;
+    }
+
+    let event_type = event >> 24;
+    let event_btn = event & 0xFFFFFF;
+
+    let event = ButtonEvent::new(event_type, event_btn);
+
+    if let Ok(event) = event {
+        return Some(event);
+    }
+    None
+}
+
+fn run<F>(frame: &mut F) -> u32
+where
+    F: Component,
+    F::Msg: ReturnToC,
+{
+    frame.place(SCREEN_ADJ);
+    frame.paint();
+    fade_backlight_duration(BACKLIGHT_NORMAL as _, 500);
+
+    while button_eval().is_some() {}
```

Comment:
Is this some kind of button debounce?

Reply:
this aims to flush any pending button events. since the buttons aren't checked all the time in bootloader there is possibility that user presses a button before screen is shown and unwillingly confirms something.